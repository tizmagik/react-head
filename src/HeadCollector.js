import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Provider } from './headTagsContext';

export default class HeadCollector extends Component {
  static propTypes = {
    headTags: PropTypes.array.isRequired, // eslint-disable-line react/forbid-prop-types
    children: PropTypes.node.isRequired,
  };

  headTags = {
    add: tag => this.props.headTags.push(tag),
  };

  render() {
    return <Provider value={this.headTags}>{React.Children.only(this.props.children)}</Provider>;
  }
}
