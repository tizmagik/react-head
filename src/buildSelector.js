const buildSelector = obj =>
  Object.keys(obj)
    .map(k => `[${k}="${obj[k]}"]`)
    .join('');

export default buildSelector;
