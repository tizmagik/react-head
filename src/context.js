import * as React from 'react';

export const { Consumer, Provider } = React.createContext(null);
