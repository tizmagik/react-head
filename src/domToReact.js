import React from 'react';

/**
 * inspired by https://github.com/remarkablemark/html-react-parser/blob/master/lib/dom-to-react.js
 * but a very stripped down version
 * - no support for
 *  - replace
 *  - transforming attribs (any that are needed should be hard coded here)
 *  - style props (not needed in head)
 * 
 * Will work with script tags however its NOT FOR INJECTING CODE - use this for things like <Script type="application/ld+json" dangerouslySetInnerHTML={{ __html: JSON.stringify(schemaOrg) }}/>
 */
const domToReact =(nodes=[], options={}) => {

  const library = options.library || React;
  const { createElement } = library;

  const result = []

  const len = nodes.length

  for (let i = 0; i < len; i+=1) {
    let node = nodes[i];

    let children = null;
    const props = {...node.attribs}

    switch (node.type) {
      case 'script':
      case 'style':
        // prevent text in <script> or <style> from being escaped
        // https://reactjs.org/docs/dom-elements.html#dangerouslysetinnerhtml
        if (node.children[0]) {
          props.dangerouslySetInnerHTML = {
            __html: node.children[0].data
          };
        }
        break;

      case 'tag':
        if (node.children && node.children.length) {
          // continue recursion of creating React elements (if applicable)
          children = domToReact(node.children, options);
        }
        break;

      // skip all other cases (e.g., comment, text)
      default:
        node = null
    }
    
    if (node) {
      // set "key" prop for sibling elements
      // https://fb.me/react-warning-keys
      if (len > 1) {
        props.key = i;
      }

      result.push(createElement(node.name, props, children));
    }
  }

  return result

}
export default domToReact