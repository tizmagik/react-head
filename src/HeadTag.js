import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import buildSelector from './buildSelector';

export default class HeadTag extends Component {
  static contextTypes = {
    reactHeadTags: PropTypes.object,
  };

  static propTypes = {
    tag: PropTypes.string,
  };

  static defaultProps = {
    tag: 'meta',
  };

  state = {
    canUseDOM: false,
  };

  componentDidMount() {
    // eslint-disable-next-line react/no-did-mount-set-state
    this.setState({ canUseDOM: true });

    const { tag, children, ...rest } = this.props; // eslint-disable-line react/prop-types
    const ssrTags = document.head.querySelector(`${tag}${buildSelector(rest)}[data-rh=""]`);

    /* istanbul ignore else */
    if (ssrTags) {
      ssrTags.remove();
    }
  }

  render() {
    const { tag: Tag, ...rest } = this.props;

    if (this.state.canUseDOM) {
      const Comp = <Tag {...rest} />;
      return ReactDOM.createPortal(Comp, document.head);
    }

    // on client we don't require HeadCollector
    if (this.context.reactHeadTags) {
      const ServerComp = <Tag data-rh="" {...rest} />;
      this.context.reactHeadTags.add(ServerComp);
    }

    return null;
  }
}
