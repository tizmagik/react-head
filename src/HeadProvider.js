import * as React from 'react';
import { Provider } from './context';

const cascadingTags = ['title', 'meta'];

export default class HeadProvider extends React.Component {
  indices = new Map();

  state = {
    addClientTag: (tag, name) => {
      // consider only cascading tags
      if (cascadingTags.indexOf(tag) !== -1) {
        this.setState(state => {
          const names = state[tag] || [];
          return { [tag]: [...names, name] };
        });
        // track indices synchronously
        const { indices } = this;
        const index = indices.has(tag) ? indices.get(tag) + 1 : 0;
        indices.set(tag, index);
        return index;
      }
      return -1;
    },

    shouldRenderTag: (tag, index) => {
      if (cascadingTags.indexOf(tag) !== -1) {
        const names = this.state[tag]; // eslint-disable-line react/destructuring-assignment
        // check if the tag is the last one of similar
        return names && names.lastIndexOf(names[index]) === index;
      }
      return true;
    },

    removeClientTag: (tag, index) => {
      this.setState(state => {
        const names = state[tag];
        if (names) {
          names[index] = null;
          return { [tag]: names };
        }
        return null;
      });
    },

    addServerTag: tagNode => {
      const { headTags = [] } = this.props;
      // tweak only cascading tags
      if (cascadingTags.indexOf(tagNode.type) !== -1) {
        const index = headTags.findIndex(prev => {
          const prevName = prev.props.name || prev.props.property;
          const nextName = tagNode.props.name || tagNode.props.property;
          return prev.type === tagNode.type && prevName === nextName;
        });
        if (index !== -1) {
          headTags.splice(index, 1);
        }
      }
      headTags.push(tagNode);
    },
  };

  componentDidMount() {
    const ssrTags = document.head.querySelectorAll(`[data-rh=""]`);
    // `forEach` on `NodeList` is not supported in Googlebot, so use a workaround
    Array.prototype.forEach.call(ssrTags, ssrTag =>
      ssrTag.parentNode.removeChild(ssrTag)
    );
  }

  render() {
    const { headTags, children } = this.props;

    if (typeof window === 'undefined' && Array.isArray(headTags) === false) {
      throw Error(
        'headTags array should be passed to <HeadProvider /> in node'
      );
    }

    return <Provider value={this.state}>{children}</Provider>;
  }
}
