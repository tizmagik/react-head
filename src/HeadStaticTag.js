import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { Consumer } from './context';
import domToReact from "./domToReact"

export default class HeadStaticTag extends React.Component {
  state = {
    canUseDOM: false,
  };

  componentDidMount() {
    this.setState({ canUseDOM: true });
  }

  render() {
    const { canUseDOM } = this.state;

    return (
      <Consumer>
        {api => {
          if (api == null) {
            throw Error('<HeadProvider /> should be in the tree');
          }

          const { json } = this.props
          const components = domToReact(json)
          if (canUseDOM) {
            return ReactDOM.createPortal(components, document.head);
          }
          api.addServerComponents(components);
          return null;
        }}
      </Consumer>
    );
  }
}
