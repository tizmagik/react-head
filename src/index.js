import React from 'react';
import HeadTag from './HeadTag';

export const Title = props => <HeadTag tag="title" {...props} />;

export const Style = props => <HeadTag tag="style" {...props} />;

export const Meta = props => <HeadTag tag="meta" {...props} />;

export const Link = props => <HeadTag tag="link" {...props} />;

export { default as HeadCollector } from './HeadCollector';

export default HeadTag;
