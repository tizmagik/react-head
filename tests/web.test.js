import * as React from 'react';
import * as plug from 'react-powerplug';
import TestRenderer from 'react-test-renderer';
import './ReactDOMMock';
import { HeadProvider, Title, Style, Meta, Link } from '../src';

test('renders into document.head portal', () => {
  const renderer = TestRenderer.create(
    <HeadProvider>
      <div>
        Yes render
        <Title>Test title</Title>
        <Style>{`body {}`}</Style>
        <Link href="index.css" />
        <Meta charSet="utf-8" />
      </div>
    </HeadProvider>
  );
  expect(renderer.toJSON()).toMatchSnapshot();
});

test('renders only the last title', () => {
  const renderer = TestRenderer.create(
    <HeadProvider>
      <div>
        <Title>Title 1</Title>
      </div>
      <div>
        <Title>Title 2</Title>
      </div>
      <div>
        <Title>Title 3</Title>
      </div>
    </HeadProvider>
  );
  expect(renderer.toJSON()).toMatchSnapshot();
});

test('mounts and unmounts title', () => {
  const renderer = TestRenderer.create(
    <HeadProvider>
      <Title>Static</Title>
      <plug.Toggle initial={false}>
        {title => (
          <>
            {title.on && <Title>Dynamic</Title>}
            <button onClick={title.toggle}>toggle</button>
          </>
        )}
      </plug.Toggle>
    </HeadProvider>
  );
  expect(renderer.toJSON()).toMatchSnapshot();
  // mount
  renderer.root.findByType('button').props.onClick();
  expect(renderer.toJSON()).toMatchSnapshot();
  // unmount
  renderer.root.findByType('button').props.onClick();
  expect(renderer.toJSON()).toMatchSnapshot();
});

test('switches between titles', () => {
  const renderer = TestRenderer.create(
    <HeadProvider>
      <Title>Static</Title>
      <plug.Value initial={null}>
        {title => (
          <>
            {title.value === 0 && <Title>Title 1</Title>}
            <button onClick={() => title.set(0)}>0</button>
            {title.value === 1 && <Title>Title 2</Title>}
            <button onClick={() => title.set(1)}>1</button>
          </>
        )}
      </plug.Value>
    </HeadProvider>
  );
  expect(renderer.toJSON()).toMatchSnapshot();
  // enable 0
  renderer.root.findAllByType('button')[0].props.onClick();
  expect(renderer.toJSON()).toMatchSnapshot();
  // switch to 1
  renderer.root.findAllByType('button')[1].props.onClick();
  expect(renderer.toJSON()).toMatchSnapshot();
  // switch to 0
  renderer.root.findAllByType('button')[0].props.onClick();
  expect(renderer.toJSON()).toMatchSnapshot();
});

test('renders only the last meta with the same name', () => {
  const renderer = TestRenderer.create(
    <HeadProvider>
      <Meta>Static 1</Meta>
      <Meta name="name1">Static 2</Meta>
      <plug.Toggle initial={false}>
        {meta => (
          <>
            {meta.on && <Meta name="name1">Dynamic 1</Meta>}
            <button onClick={meta.toggle}>toggle</button>
          </>
        )}
      </plug.Toggle>
      <plug.Toggle initial={false}>
        {meta => (
          <>
            {meta.on && <Meta>Dynamic 2</Meta>}
            <button onClick={meta.toggle}>toggle</button>
          </>
        )}
      </plug.Toggle>
    </HeadProvider>
  );
  expect(renderer.toJSON()).toMatchSnapshot();
  // mount first
  renderer.root.findAllByType('button')[0].props.onClick();
  expect(renderer.toJSON()).toMatchSnapshot();
  // mount second
  renderer.root.findAllByType('button')[1].props.onClick();
  expect(renderer.toJSON()).toMatchSnapshot();
  // unmount second
  renderer.root.findAllByType('button')[1].props.onClick();
  expect(renderer.toJSON()).toMatchSnapshot();
  // unmount first
  renderer.root.findAllByType('button')[0].props.onClick();
  expect(renderer.toJSON()).toMatchSnapshot();
});

test('renders only last meta with the same property', () => {
  const renderer = TestRenderer.create(
    <HeadProvider>
      <Meta property="name1">Meta 1</Meta>
      <Meta property="name1">Meta 2</Meta>
      <Meta property="name2">Meta 3</Meta>
      <Meta name="name2">Meta 4</Meta>
      <Meta name="name3">Meta 5</Meta>
      <Meta property="name3">Meta 6</Meta>
    </HeadProvider>
  );
  expect(renderer.toJSON()).toMatchSnapshot();
});

test('throws error if head tag is rendered without HeadProvider', () => {
  const errorFn = jest.spyOn(console, 'error').mockImplementation(() => {});
  expect(() => {
    TestRenderer.create(<Style>{`body {}`}</Style>);
  }).toThrowError(/<HeadProvider \/> should be in the tree/);
  errorFn.mockRestore();
});
