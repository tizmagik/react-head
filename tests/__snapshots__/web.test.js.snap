// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mounts and unmounts title 1`] = `
Array [
  <portal-head>
    <title>
      Static
    </title>
  </portal-head>,
  <button
    onClick={[Function]}
  >
    toggle
  </button>,
]
`;

exports[`mounts and unmounts title 2`] = `
Array [
  <portal-head>
    <title>
      Dynamic
    </title>
  </portal-head>,
  <button
    onClick={[Function]}
  >
    toggle
  </button>,
]
`;

exports[`mounts and unmounts title 3`] = `
Array [
  <portal-head>
    <title>
      Static
    </title>
  </portal-head>,
  <button
    onClick={[Function]}
  >
    toggle
  </button>,
]
`;

exports[`renders into document.head portal 1`] = `
<div>
  Yes render
  <portal-head>
    <title>
      Test title
    </title>
  </portal-head>
  <portal-head>
    <style>
      body {}
    </style>
  </portal-head>
  <portal-head>
    <link
      href="index.css"
    />
  </portal-head>
  <portal-head>
    <meta
      charSet="utf-8"
    />
  </portal-head>
</div>
`;

exports[`renders only last meta with the same property 1`] = `
Array [
  <portal-head>
    <meta
      property="name1"
    >
      Meta 2
    </meta>
  </portal-head>,
  <portal-head>
    <meta
      name="name2"
    >
      Meta 4
    </meta>
  </portal-head>,
  <portal-head>
    <meta
      property="name3"
    >
      Meta 6
    </meta>
  </portal-head>,
]
`;

exports[`renders only the last meta with the same name 1`] = `
Array [
  <portal-head>
    <meta>
      Static 1
    </meta>
  </portal-head>,
  <portal-head>
    <meta
      name="name1"
    >
      Static 2
    </meta>
  </portal-head>,
  <button
    onClick={[Function]}
  >
    toggle
  </button>,
  <button
    onClick={[Function]}
  >
    toggle
  </button>,
]
`;

exports[`renders only the last meta with the same name 2`] = `
Array [
  <portal-head>
    <meta>
      Static 1
    </meta>
  </portal-head>,
  <portal-head>
    <meta
      name="name1"
    >
      Dynamic 1
    </meta>
  </portal-head>,
  <button
    onClick={[Function]}
  >
    toggle
  </button>,
  <button
    onClick={[Function]}
  >
    toggle
  </button>,
]
`;

exports[`renders only the last meta with the same name 3`] = `
Array [
  <portal-head>
    <meta
      name="name1"
    >
      Dynamic 1
    </meta>
  </portal-head>,
  <button
    onClick={[Function]}
  >
    toggle
  </button>,
  <portal-head>
    <meta>
      Dynamic 2
    </meta>
  </portal-head>,
  <button
    onClick={[Function]}
  >
    toggle
  </button>,
]
`;

exports[`renders only the last meta with the same name 4`] = `
Array [
  <portal-head>
    <meta>
      Static 1
    </meta>
  </portal-head>,
  <portal-head>
    <meta
      name="name1"
    >
      Dynamic 1
    </meta>
  </portal-head>,
  <button
    onClick={[Function]}
  >
    toggle
  </button>,
  <button
    onClick={[Function]}
  >
    toggle
  </button>,
]
`;

exports[`renders only the last meta with the same name 5`] = `
Array [
  <portal-head>
    <meta>
      Static 1
    </meta>
  </portal-head>,
  <portal-head>
    <meta
      name="name1"
    >
      Static 2
    </meta>
  </portal-head>,
  <button
    onClick={[Function]}
  >
    toggle
  </button>,
  <button
    onClick={[Function]}
  >
    toggle
  </button>,
]
`;

exports[`renders only the last title 1`] = `
Array [
  <div />,
  <div />,
  <div>
    <portal-head>
      <title>
        Title 3
      </title>
    </portal-head>
  </div>,
]
`;

exports[`switches between titles 1`] = `
Array [
  <portal-head>
    <title>
      Static
    </title>
  </portal-head>,
  <button
    onClick={[Function]}
  >
    0
  </button>,
  <button
    onClick={[Function]}
  >
    1
  </button>,
]
`;

exports[`switches between titles 2`] = `
Array [
  <portal-head>
    <title>
      Title 1
    </title>
  </portal-head>,
  <button
    onClick={[Function]}
  >
    0
  </button>,
  <button
    onClick={[Function]}
  >
    1
  </button>,
]
`;

exports[`switches between titles 3`] = `
Array [
  <button
    onClick={[Function]}
  >
    0
  </button>,
  <portal-head>
    <title>
      Title 2
    </title>
  </portal-head>,
  <button
    onClick={[Function]}
  >
    1
  </button>,
]
`;

exports[`switches between titles 4`] = `
Array [
  <portal-head>
    <title>
      Title 1
    </title>
  </portal-head>,
  <button
    onClick={[Function]}
  >
    0
  </button>,
  <button
    onClick={[Function]}
  >
    1
  </button>,
]
`;
